src/hyperv/vmbus/hyperv.c

```diff
@@ -131,22 +139,80 @@ hypercall_signal_event(bus_addr_t monprm_paddr)
         HYPERCALL_SIGNAL_EVENT, monprm_paddr, 0);
 }
 
+static
+u64 read_hyperv_timer_tsc(void)
+{
+    struct hyperv_reftsc *tsc_ref = hyperv_info.hyperv_ref_tsc.tsc_ref;
+    u32 seq;
+
+    while ((seq = atomic_load_acq32(&tsc_ref->tsc_seq)) != 0) {
+        u64 disc, ret, tsc;
+        u64 scale = tsc_ref->tsc_scale;
+        s64 ofs = tsc_ref->tsc_ofs;
+
+        /* rdtsc_precise contains a load fence */
+        tsc = rdtsc_precise();
```

Comment:
Is rdtsc_precise absolutely necessary? A while back we had this barrier in our tsc read routine, and in our first few ftrace runs we discovered that it was causing a huge performance hit, apparently due to cpuid causing a full vm exit. I don't know if this is the case for hyperv, I guess it would be trivial to measure (we were just running webg / ab tests).

(Also, so I've read anyway, cpuid doesn't just act as a load fence, but it serializes all instruction execution around it.)